src/lib/mina_ledger/ledger.ml

```diff
@@ -419,6 +419,166 @@ let apply_initial_ledger_state : t -> init_state -> unit =
       in
       create_new_account_exn t account_id account')
 
+let%test_unit "tokens test" =
+  let open Mina_transaction_logic.For_tests in
+  let keypairs = Quickcheck.random_value (Init_ledger.gen ()) in
+  let get ledger pk token =
+    match
+      Ledger_inner.get_or_create ledger (Account_id.create pk token)
+      |> Or_error.ok_exn
+    with
+    | `Added, _, _ ->
+        failwith "Account did not exist"
+    | `Existed, a, _ ->
+        a
+  in
+  let mk_parties_transaction ledger other_parties : Parties.t =
+    let fee_payer : Party.Fee_payer.t =
+      let kp, _ = keypairs.(0) in
+      let pk = Public_key.compress kp.public_key in
+      let _, ({ nonce; _ } : Account.t), _ =
+        Ledger_inner.get_or_create ledger
+          (Account_id.create pk Token_id.default)
+        |> Or_error.ok_exn
+      in
+      { data =
+          { body =
+              { update = Party.Update.noop
+              ; public_key = pk
+              ; token_id = ()
+              ; balance_change = Currency.Fee.of_int 7
+              ; increment_nonce = ()
+              ; events = []
+              ; sequence_events = []
+              ; call_data =
+                  (* TODO: Not sure why this is a type error. *)
+                  Obj.magic Pickles.Impls.Step.Field.Constant.zero
+              ; call_depth = 0
+              ; protocol_state = Snapp_predicate.Protocol_state.accept
+              ; use_full_commitment = ()
+              }
+          ; predicate = nonce
+          ; caller = ()
+          }
+      ; authorization = Signature.dummy
+      }
+    in
+    { fee_payer
+    ; memo = Signed_command_memo.dummy
+    ; other_parties =
+        other_parties |> Parties.Call_forest.add_callers'
+        |> Parties.Call_forest.accumulate_hashes_predicated
+        |> Parties.Call_forest.map ~f:(fun p : Party.t ->
+               { data = p; authorization = Signature Signature.dummy })
+    }
+  in
+  let main (ledger : t) =
+    let execute_parties_transaction
+        (parties : (Party.Predicated.Wire.t, unit) Parties.Call_forest.t) : unit
+        =
+      let _res =
+        apply_parties_unchecked ~constraint_constants ~state_view:view ledger
+          (mk_parties_transaction ledger parties)
+        |> Or_error.ok_exn
+      in
+      ()
+    in
+    let party caller kp token_id balance_change : Party.Predicated.Wire.t =
+      { body =
+          { update = Party.Update.noop
+          ; public_key = Public_key.compress kp.Keypair.public_key
+          ; token_id
+          ; balance_change =
+              Currency.Amount.Signed.create
+                ~magnitude:(Currency.Amount.of_int (Int.abs balance_change))
+                ~sgn:(if Int.is_negative balance_change then Sgn.Neg else Pos)
+          ; increment_nonce = true
+          ; events = []
+          ; sequence_events = []
+          ; call_data =
+              (* TODO: Not sure why this is a type error. *)
+              Obj.magic Pickles.Impls.Step.Field.Constant.zero
```

Comment:
This didn't typecheck for some reason

Reply:
This is probably because of the `no-implicit-deps` stanza that was added to `dune-project`. It's indicative of a code smell: we access the field type and its members through a large number of different modules, where we ought to have one canonical module and use that directly.